ntp-proto/src/filter.rs

```diff
@@ -0,0 +1,315 @@
+// An implementation of the NTP clock filter algorithm, as described by
+//
+//      https://datatracker.ietf.org/doc/html/rfc5905#page-37
+//
+// Specifically this is a rust implementation of the `clock_filter()` routine,
+// described in the appendix
+//
+//      https://datatracker.ietf.org/doc/html/rfc5905#appendix-A.5.2
+
+use crate::{packet::NtpLeapIndicator, NtpDuration, NtpTimestamp};
+
+/// frequency tolerance (15 ppm)
+const ONE_OVER_PHI: i64 = 15_000_000;
+
+/// spike gate (clock filter)
+const SGATE: f64 = 3.0;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+struct FilterTuple {
+    offset: NtpDuration,
+    delay: NtpDuration,
+    dispersion: NtpDuration,
+    time: NtpTimestamp,
+}
+
+impl FilterTuple {
+    fn dummy() -> Self {
+        Self {
+            offset: NtpDuration::default(),
+            delay: NtpDuration::MAXDISP,
+            dispersion: NtpDuration::MAXDISP,
+            time: NtpTimestamp::default(),
+        }
+    }
+
+    fn is_dummy(self) -> bool {
+        self == Self::dummy()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct ClockFilterContents {
+    register: [FilterTuple; 8],
+}
+
+impl ClockFilterContents {
+    #[allow(dead_code)]
+    fn new() -> Self {
+        Self {
+            register: [FilterTuple::dummy(); 8],
+        }
+    }
+
+    fn shift_and_insert(&mut self, new_tuple: FilterTuple, dispersion_correction: NtpDuration) {
+        let mut current = new_tuple;
+
+        for tuple in self.register.iter_mut() {
+            tuple.dispersion += dispersion_correction;
+
+            std::mem::swap(&mut current, tuple);
+        }
+    }
+
+    fn sort_by_delay(&mut self) {
+        self.register.sort_by(|t1, t2| {
+            t1.delay
+                .partial_cmp(&t2.delay)
+                .unwrap_or_else(|| panic!("got a NaN"))
+        });
+    }
```

Comment:
can we just ignore `NaN` and pick an ordering here? I don't see downsides

Reply:
Maybe consider https://crates.io/crates/ordered-float?